packages/flutter_tools/lib/src/android/gradle.dart

```diff
@@ -211,445 +192,285 @@ void createSettingsAarGradle(Directory androidDirectory) {
+    command.add('-Dbuild-plugins-as-aars=true');
+    // Don't use settings.gradle from the current project since it includes the plugins as subprojects.
+    command.add('--settings-file=settings_aar.gradle');
   }
-  final String androidPluginVersion = pluginMatches.first.group(1);
-  return getGradleVersionFor(androidPluginVersion);
-}
+  command.add(assembleTask);
 
-/// Overwrite local.properties in the specified Flutter project's Android
-/// sub-project, if needed.
-///
-/// If [requireAndroidSdk] is true (the default) and no Android SDK is found,
-/// this will fail with a [ToolExit].
-void updateLocalProperties({
-  @required FlutterProject project,
-  BuildInfo buildInfo,
-  bool requireAndroidSdk = true,
-}) {
-  if (requireAndroidSdk) {
-    _exitIfNoAndroidSdk();
+  final Stopwatch sw = Stopwatch()..start();
+  int exitCode = 1;
+  GradleHandledError detectedGradleError;
+  String detectedGradleErrorLine;
+  try {
+    exitCode = await processUtils.stream(
+      command,
+      workingDirectory: project.android.hostAppGradleRoot.path,
+      allowReentrantFlutter: true,
+      environment: gradleEnv,
+      mapFunction: (String line) {
+        // This message was removed from 1P plugins,
+        // but older plugin versions still display this message.
+        if (androidXPluginWarningRegex.hasMatch(line)) {
+          // Don't pipe.
+          return null;
+        }
+        if (detectedGradleError == null) {
+          for (final GradleHandledError gradleError in gradleErrors) {
+            if (gradleError.test(line)) {
+              detectedGradleErrorLine = line;
+              detectedGradleError = gradleError;
+            }
+          }
+        }
+        // Pipe stdout/sterr from Gradle.
+        return line;
+      },
+    );
+  } finally {
+    status.stop();
   }
 
-  final File localProperties = project.android.localPropertiesFile;
-  bool changed = false;
-
-  SettingsFile settings;
-  if (localProperties.existsSync()) {
-    settings = SettingsFile.parseFromFile(localProperties);
-  } else {
-    settings = SettingsFile();
-    changed = true;
+  if (exitCode != 0 && detectedGradleError == null) {
+    BuildEvent('unknown-failure').send();
+    throwToolExit(
+      'Gradle task $assembleTask failed with exit code $exitCode',
+      exitCode: exitCode,
+    );
   }
 
-  void changeIfNecessary(String key, String value) {
-    if (settings.values[key] != value) {
-      if (value == null) {
-        settings.values.remove(key);
-      } else {
-        settings.values[key] = value;
-      }
-      changed = true;
+  if (exitCode != 0 && detectedGradleError != null) {
+    final GradleBuildStatus status = await detectedGradleError.handler(
+      line: detectedGradleErrorLine,
+      project: project,
+      usesAndroidX: usesAndroidX,
+      shouldBuildPluginAsAar: shouldBuildPluginAsAar,
+    );
+    if (status == GradleBuildStatus.RETRY && retries >= 1) {
+      await buildGradleApp(
```

Comment:
Would it be interesting to log an analytics event for these retries?

Reply:
The change I made now logs the retries as well, so it's easy to know if the retry fixed the issue.